common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
@arvindsv 
dude, don't you need sleep ? :-D

On material identity:
Maybe I am wrong, but submoduleFolder is not included in the identity is problematic. Given following material configs:

```xml
<pipeline name="p1">
      <materials>
        <git url="https://the-repo" materialName="the-repo" submoduleFolder="foo" />
     <materials>
</pipeline>

<pipeline name="p2">
      <materials>
        <git url="https://the-repo" materialName="the-repo"  submoduleFolder="bar" />
     <materials>
</pipeline>
```
How do we guarantee p1 build get submoduleFolder "foo" and p2 build get submoduleFolder "bar"?